### ISSUE TITLE
Mail Merge in Gmail doesn't work

### ISSUE THREAD

**Issue by audioa86** [googleworkspace/solutions#155]
The first 2 times I tried Mail Merge in Gmail by following the directions [here](https://www.youtube.com/watch?v=YTeV4i1W3Zo&list=PLU8ezI8GYqs4YntFNP9jf_rrZ0vJLSW2X&index=4) it worked. However, beginning yesterday, it has stopped working. The markers in the email draft correspond to the column headers in the Google Sheet. (I copied the header names and pasted them in the email draft with the double brackets around them, like this: {{First name}}. The emails will send with the email draft to the recipients listed in the Google Sheet, but none of the fields within the email body will fill in. Instead there are blanks where they are missing. This seems like a glitch in the system. What can I do to fix this?

**Comment by drlube** [comment]
I tried the sheet/script for the first time today, and am experiencing the same issue.

**Comment by fizzatbeyond** [comment]
My wife encountered the same problem.  She had previously used to generate approximately 500 emails without an issue however this time when generating about 1200 emails the data was blank.  She broke it into two smaller groupings and they both went through with all the data as expected.

I suspect the problem has to do with some maximum number of rows or amount of data being processed.

**Comment by audioa86** [comment]
Unfortunately, it doesn't have to do with maximum rows of data for me. I was demonstrating how to use the tool by using only 2 rows of test data (with real email addresses) when this stopped working properly.

**Comment by vanowarna** [comment]
The same occurred to me with just 2 rows. Emails were sent separately and the {{ }} in the subject was Changed as expected. But the body was blank. Is there a fix?

`// loop through all the rows of data
  obj.forEach(function(row, rowIdx){
    // only send emails is email_sent cell is blank and not hidden by filter`

Line 81 : What is rowIdx that hasn't mentioned anywhere. I suspect that's the issue. Any fixes?

**Comment by CampbellDuke** [comment]
I'm having the same issue - using just one line as a test.

**Comment by vanowarna** [comment]
### Found a solution. 
It worked for me. When pasting the names of the columns in the email body. **Paste them as plain text(ctrl+shift+v)**. Otherwise the formatting interrupts the algorithm.

**Comment by CampbellDuke** [comment]
Thanks vanowarna. I just tried clearing all formatting and sending a test to myself again. Same issue. Then I just typed in my name and email address after clearing formatting in all cells. Ditto.

I still think it's useful, it just looks like I won't be able to personalize things.

**Comment by gsaxer** [comment]
I have this same issue today.
The part which is very confusing is that if I look at the email using "Show Original" I CAN SEE THE INSERTED TEXT just fine. The HTML version as well as the info at the bottom does NOT have the text though.

For example, the draft email has

{{Discount code}} 4
{{First name}} 5

in it.

In the non-HTML part of the Show Original I see

xx 4
Gary 5

(Those are the right test values in the spreadsheet.)

However, in the area with

Content-Type: text/html; charset="UTF-8"
Content-Transfer-Encoding: quoted-printable

I can see

<div><spa=
n style=3D"color:rgb(0,0,0);font-family:Roboto,RobotoDraft,Helvetica,Arial,=
sans-serif;font-size:13px;white-space:pre-wrap"> 4</span></div><div><span s=
tyle=3D"color:rgb(0,0,0);font-family:Roboto,RobotoDraft,Helvetica,Arial,san=
s-serif;font-size:13px;white-space:pre-wrap"> 5</span></div>

So somehow the actual text is getting dropped.

**Comment by vanowarna** [comment]
> I have this same issue today.
> The part which is very confusing is that if I look at the email using "Show Original" I CAN SEE THE INSERTED TEXT just fine. The HTML version as well as the info at the bottom does NOT have the text though.
> 
> For example, the draft email has
> 
> {{Discount code}} 4
> {{First name}} 5
> 
> in it.
> 
> In the non-HTML part of the Show Original I see
> 
> xx 4
> Gary 5
> 
> (Those are the right test values in the spreadsheet.)
> 
> However, in the area with
> 
> Content-Type: text/html; charset="UTF-8"
> Content-Transfer-Encoding: quoted-printable
> 
> I can see
> 
> 4
> 5
> So somehow the actual text is getting dropped.

Did you try this  https://github.com/googleworkspace/solutions/issues/155#issuecomment-764383214

**Comment by gsaxer** [comment]
Yes, indeed I made sure that there was no formatting of the names in {{}}.
Note in my example there's a 4 and 5. That's because I had multiple tests. One was an exact copy from the spreadsheet into the draft email and that included the name being in bold and another was it as plain text.

I also see that the quoted-printable stuff is getting "processed" by Github's comment section. Ugh. Let me see if I can post it again:

"<div><spa=
n style=3D"color:rgb(0,0,0);font-family:Roboto,RobotoDraft,Helvetica,Arial,=
sans-serif;font-size:13px;white-space:pre-wrap"> 4</span></div><div><span s=
tyle=3D"color:rgb(0,0,0);font-family:Roboto,RobotoDraft,Helvetica,Arial,san=
s-serif;font-size:13px;white-space:pre-wrap"> 5</span></div>"

(In the preview this info is now showing. This is the relevant chunk of stuff in the email from which the displayed email is generated. It's at the end of Show Original.)

This HTML stuff shows the 4 and 5 but not the values from the spreadsheet. As mentioned before, in the text of the Show Original the values from the spreadsheet are THERE. Something is happening when the "conversion" to HTML is taking place.

**Comment by gsaxer** [comment]
Update - The account I'm using to do this Mail Merge is a "G Suite" version.
I just tried this same thing in my personal email account and the values from the spreadsheet now appear in the sent emails!

Significantly, in the quoted-printable section of Show Original there's none of the style or other stuff, it's just

"
<br>z 4<br>Gary 5<br>
![image](https://user-images.githubusercontent.com/3157185/105617081-396cba00-5d91-11eb-8dfc-cb0257a68b6d.png)
"

which are the values from the spreadsheet and the number next to the value.
(The above is the actual code in the quotes and then an image snipped from the screen because I can't get this editor to drop processing it as HTML.)

**Comment by gsaxer** [comment]
Another update:
Based on this "plain text" concept I decided to select the entire draft email and then "cut" it so the email body was empty. Then did a ctrl-shift-v (paste as plain text) for the draft email. A little "cleaning up" was needed - mostly putting in come extra "return" items at the places where an empty line was desired after a paragraph.

THIS WORKED - even in the GSuite version.

So, getting rid of all "formatting" in the email did the trick. I note that the sent email automatically had any web "links" converted to a clickable link (which is fine) but otherwise all the other formatting was "plain" in the resulting email. However, the values from the spreadsheet were put in the right place where {{Discount code}} etc. were.

Success!

**Comment by brew-guy** [comment]
I had the same problem and found that it was case sensitive so after making sure my {{}} tags text had exactly same case as sheet header text it worked.

**Comment by rupayvidya** [comment]
Thank You @gsaxer 
You made my Day.
I followed your steps by copying the complete mail and pasting it as plain text. It worked.

Thank you very much.

**Comment by timhughes-dev** [comment]
My colleague is having the same problem with fields appearing blank in the emails sent.

The script is working fine for me.  I've tried adding formatting and it still works fine.  Does anyone have any other suggestions on what else my colleague should try to solve this?

**Comment by yotiko** [comment]
thnx folks for the solution, has somebody found a solution for the formating problem?
i would like to make some text to appear **bold**

**Comment by jyotingautam** [comment]
@yotiko  The summary of the solution is:
1. First do a plain-text copy of the email body as @gsaxer  explains (by cutting your entire email body with cntrl-X  (or cmd-x on Macbook) and then pasting unformatted using cntrl-shift-v (or cmd-shift-v on Macbook)
2.For formatting the email, next step is to select the text to format (but don't touch the Mail-merge variables!!) and format the the text using bold, italic etc. Don't touch the mail-merge variables when re-formatting!

This worked fine for me. @gsaxer  was a lifesaver - thanks!

**Comment by MsTsam000** [comment]
@gsaxer you are life saver...spent almost 4 hrs trying to find the solution by scouring youtube. So I am confirming that changing the text formatting by converting the body of the email to " Plain Text" via the (Crt+Shift+V) was the solution. 
I was never a member of this forum...but found it via google and I had to sign up to show my gratitude!

So Thanks again @gsaxer !

**Comment by Fidle123** [comment]
Thanks, Vanowarna

Great help. 

It worked for me too, with this method as well. By removing the formatting and typing inside the curly brackets exactly as it is example {{First name}}

**Comment by Fidle123** [comment]
> The same occurred to me with just 2 rows. Emails were sent separately and the {{ }} in the subject was Changed as expected. But the body was blank. Is there a fix?
> 
> `// loop through all the rows of data obj.forEach(function(row, rowIdx){ // only send emails is email_sent cell is blank and not hidden by filter`
> 
> Line 81 : What is rowIdx that hasn't mentioned anywhere. I suspect that's the issue. Any fixes?

Thanks, Vanowarna

Great help.

It worked for me too, with this method as well. By removing the formatting and typing inside the curly brackets exactly as it is example {{First name}}

**Comment by jpoehnelt** [comment]
Closing this issue as the repository is archived.
